transport/resolver.go

```diff
@@ -0,0 +1,23 @@
+// Copyright 2019 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transport
+
+import (
+	"net"
+	"net/url"
+)
+
+// Resolver is a function that resolves a url to a list of IP addresses.
+type Resolver func(url *url.URL) ([]net.Addr, error)
```

Comment:
I think it would be helpful to show how it might be used in the future. To me it should cover the following two scenarios (feel free to correct me if it's not the case):

1. Pass a resolver instance to Go's standard [`net/Dialer.Resolver`](https://pkg.go.dev/net#Dialer)
2. How to handle DNS requests by a resolver instance in [`PacketProxy`](https://github.com/Jigsaw-Code/outline-sdk/blob/main/network/packet_proxy.go)

We can add it to the documentation (it's just an example here, you can write in your preferred way):

```
// Resolver is a function that resolves a url to a list of IP addresses.
// We provided factory methods to create a Resolver, for example, NewDNSResolver(dnsaddr), NewDNSOverHTTPResolver(httpaddr).
// To use the resolver in Go, you can:
//
//     doh := NewDNSOverHTTPResolver("8.8.8.8:80")
//     r := &net.Resolver{
//       Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
//         ?? use doh to create a net.Conn ??
//       },
//     }
//     dialer := &net.Dialer{Resolver: r}
// 
// To use the resolver to handle UDP traffic:
//
//     doh := NewDNSOverHTTPResolver("8.8.8.8:80")
//     nonDnsTrafficProxy := ...
//     pp := NewDNSPacketProxy(doh, nonDnsTrafficProxy)
//
// ...
```

I feel it's helpful to think about use cases when designing an interface. And it also helps reviewers to understand the context as well.

Reply:
+1 to clarifying the use cases.

We need a StreamDialer that resolves the destination in a resilient way. Perhaps we should look into what that looks like.

We will most certainly need to implement Happy Eyeballs in some way.